File: itk/agents/go/v03/main.go
package main

import (
	"context"
	"encoding/base64"
	"flag"
	"fmt"
	"log"
	"net"
	"net/http"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"

	"itk/agents/go/v03/pb"

	"github.com/a2aproject/a2a-go/a2a"
	"github.com/a2aproject/a2a-go/a2aclient"
	"github.com/a2aproject/a2a-go/a2aclient/agentcard"
	"github.com/a2aproject/a2a-go/a2asrv"
	"github.com/a2aproject/a2a-go/a2asrv/eventqueue"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/protobuf/proto"

	"github.com/a2aproject/a2a-go/a2agrpc"
	"golang.org/x/sync/errgroup"
)

type V03AgentExecutor struct {
}

func (e *V03AgentExecutor) Execute(ctx context.Context, reqCtx *a2asrv.RequestContext, queue eventqueue.Queue) error {
	log.Printf("Executing task %s", reqCtx.Message.ID)

	// 1. Extract Instruction from message parts
	var instruction pb.Instruction
	found := false
	for _, part := range reqCtx.Message.Parts {
		if filePart, ok := part.(a2a.FilePart); ok {
			if fileBytes, ok := filePart.File.(a2a.FileBytes); ok {
				rawBytes, err := base64.StdEncoding.DecodeString(fileBytes.Bytes)
				if err != nil {
					log.Printf("Failed to decode base64 bytes: %v", err)
					continue
				}
				if err := proto.Unmarshal(rawBytes, &instruction); err == nil {
					found = true
					break
				}
			}
		}
	}

	if !found {
		errMsg := "Error: No valid Instruction found in request."
		log.Println(errMsg)
		return queue.Write(ctx, a2a.NewMessageForTask(a2a.MessageRoleAgent, reqCtx, a2a.TextPart{Text: errMsg}))
	}

	// 2. Handle Instruction
	results, err := e.handleInstruction(ctx, reqCtx, &instruction)
	if err != nil {
		log.Printf("Error handling instruction: %v", err)
		return queue.Write(ctx, a2a.NewMessageForTask(a2a.MessageRoleAgent, reqCtx, a2a.TextPart{Text: fmt.Sprintf("Execution Error: %v", err)}))
	}

	// 3. Return response
	response := strings.Join(results, "\n")
	return queue.Write(ctx, a2a.NewMessageForTask(a2a.MessageRoleAgent, reqCtx, a2a.TextPart{Text: response}))
}

func (e *V03AgentExecutor) handleInstruction(ctx context.Context, reqCtx *a2asrv.RequestContext, inst *pb.Instruction) ([]string, error) {
	switch {
	case inst.GetCallAgent() != nil:
		call := inst.GetCallAgent()
		log.Printf("Calling agent %s via %s", call.AgentCardUri, call.Transport)

		// Resolve card and create client
		card, err := agentcard.DefaultResolver.Resolve(ctx, call.AgentCardUri)
		if err != nil {
			return nil, fmt.Errorf("failed to resolve agent card for %s: %w", call.AgentCardUri, err)
		}

		opts := []a2aclient.FactoryOption{
			a2aclient.WithGRPCTransport(grpc.WithTransportCredentials(insecure.NewCredentials())),
		}

		tp := mapTransport(call.Transport)
		log.Printf("Mapped transport: %s", tp)

		matchedInterfaces := selectInterfaces(tp, card)
		if len(matchedInterfaces) == 0 {
			return nil, fmt.Errorf("transport protocol %s is not supported by agent %s", tp, call.AgentCardUri)
		}

		// 3. Create a client with transport enforcement
		// We instantiate a client through a2aclient.CreateFromEndpoints
		// to strictly enforce the transport protocol. a2aclient.NewFromCard
		// seems to be using the first available transport if the specified one
		// is not supported.
		client, err := a2aclient.NewFromEndpoints(ctx, matchedInterfaces, opts...)
		if err != nil {
			return nil, fmt.Errorf("failed to connect to agent %s: %w", call.AgentCardUri, err)
		}

		// Wrap instruction back to a message
		wrappedMsg, err := wrapInstructionToRequest(call.Instruction)
		if err != nil {
			return nil, fmt.Errorf("failed to wrap nested instruction: %w", err)
		}

		var responses []string
		if call.Streaming {
			events := client.SendStreamingMessage(ctx, wrappedMsg)
			for ev, err := range events {
				if err != nil {
					return nil, fmt.Errorf("streaming call failed to agent %s: %w", call.AgentCardUri, err)
				}
				switch r := ev.(type) {
				case *a2a.TaskStatusUpdateEvent:
					if r.Status.Message != nil {
						for _, part := range r.Status.Message.Parts {
							if textPart, ok := part.(a2a.TextPart); ok {
								responses = append(responses, textPart.Text)
							}
						}
					}
				}
			}
		} else {
			result, err := client.SendMessage(ctx, wrappedMsg)
			if err != nil {
				return nil, fmt.Errorf("failed to send message to agent %s: %w", call.AgentCardUri, err)
			}
			switch r := result.(type) {
			case *a2a.Message:
				for _, part := range r.Parts {
					if textPart, ok := part.(a2a.TextPart); ok {
						responses = append(responses, textPart.Text)
					}
				}
			case *a2a.Task:
				if r.Status.Message != nil {
					for _, part := range r.Status.Message.Parts {
						if textPart, ok := part.(a2a.TextPart); ok {
							responses = append(responses, textPart.Text)
						}
					}
				}
				for _, msg := range r.History {
					if msg.Role == a2a.MessageRoleAgent {
						for _, part := range msg.Parts {
							if textPart, ok := part.(a2a.TextPart); ok {
								responses = append(responses, textPart.Text)
							}
						}
					}
				}
			default:
				return nil, fmt.Errorf("unexpected result type from SendMessage: %T", result)
			}
		}

		return responses, nil

	case inst.GetReturnResponse() != nil:
		return []string{inst.GetReturnResponse().Response}, nil

	case inst.GetSteps() != nil:
		var allResults []string
		for _, step := range inst.GetSteps().Instructions {
			results, err := e.handleInstruction(ctx, reqCtx, step)
			if err != nil {
				return nil, err
			}
			allResults = append(allResults, results...)
		}
		return allResults, nil

	default:
		return nil, fmt.Errorf("unknown instruction type")
	}
}

func boolPtr(b bool) *bool {
	return &b
}

func (e *V03AgentExecutor) Cancel(_ context.Context, reqCtx *a2asrv.RequestContext, _ eventqueue.Queue) error {
	log.Printf("Cancel requested for task %s", reqCtx.Message.ID)
	return nil
}

func wrapInstructionToRequest(inst *pb.Instruction) (*a2a.MessageSendParams, error) {
	instBytes, err := proto.Marshal(inst)
	if err != nil {
		return nil, err
	}
	b64Inst := base64.StdEncoding.EncodeToString(instBytes)

	msg := a2a.NewMessage(a2a.MessageRoleUser, a2a.FilePart{
		File: a2a.FileBytes{
			Bytes: b64Inst,
			FileMeta: a2a.FileMeta{
				MimeType: "application/x-protobuf",
				Name:     "instruction.bin",
			},
		},
	})

	return &a2a.MessageSendParams{
		Config: &a2a.MessageSendConfig{
			Blocking: boolPtr(true),
		},
		Message: msg,
	}, nil
}

func mapTransport(t string) a2a.TransportProtocol {
	switch strings.ToUpper(t) {
	case "GRPC":
		return a2a.TransportProtocolGRPC
	case "HTTP_JSON":
		return a2a.TransportProtocolHTTPJSON
	default:
		return a2a.TransportProtocolJSONRPC
	}
}



func selectInterfaces(tp a2a.TransportProtocol, card *a2a.AgentCard) []a2a.AgentInterface {
	var matched []a2a.AgentInterface
	prefTp := card.PreferredTransport
	if prefTp == "" {
		prefTp = a2a.TransportProtocolJSONRPC
	}
	if prefTp == tp {
		matched = append(matched, a2a.AgentInterface{
			URL:       strings.TrimSuffix(card.URL, "/"),
			Transport: tp,
			// ProtocolVersion: card.ProtocolVersion, // Wait, v0.3 AgentInterface doesn't have ProtocolVersion?
		})
	}

	for _, iface := range card.AdditionalInterfaces {
		if iface.Transport == tp {
			iface.URL = strings.TrimSuffix(iface.URL, "/")
			matched = append(matched, iface)
		}
	}
	return matched
}

var httpPort = flag.Int("httpPort", 10101, "HTTP port")
var grpcPort = flag.Int("grpcPort", 11001, "gRPC port")

func main() {
	if err := run(); err != nil {
		log.Fatalf("Server session ended with error: %v", err)
	}
}

func run() error {
	flag.Parse()

	host := "127.0.0.1"
	jsonRPCAddr := fmt.Sprintf("http://127.0.0.1:%d/jsonrpc", *httpPort)
	grpcAddr := fmt.Sprintf("127.0.0.1:%d", *grpcPort)

	skill := a2a.AgentSkill{
		ID:          "itk_v03_proto_skill",
		Name:        "ITK v03 Proto Skill",
		Description: "Handles raw byte Instruction protos in v03 subproject.",
		Tags:        []string{"proto", "v03", "itk"},
		Examples:    []string{"Roll a dice", "Call another agent"},
	}

	// go-sdk v.03 has support for JSONRPC and GRPC only
	agentCard := &a2a.AgentCard{
		Name:               "ITK v03 Agent",
		Description:        "Multi-transport agent supporting raw Instruction protos.",
		URL:                jsonRPCAddr,
		Version:            "0.3.0",
		DefaultInputModes:  []string{"text"},
		DefaultOutputModes: []string{"text"},
		Capabilities:       a2a.AgentCapabilities{Streaming: true},
		Skills:             []a2a.AgentSkill{skill},
		PreferredTransport: a2a.TransportProtocolJSONRPC,
		AdditionalInterfaces: []a2a.AgentInterface{
			{URL: jsonRPCAddr, Transport: a2a.TransportProtocolJSONRPC},
			{URL: grpcAddr, Transport: a2a.TransportProtocolGRPC},
		},
		ProtocolVersion: "0.3.0",
	}

	executor := &V03AgentExecutor{}
	requestHandler := a2asrv.NewHandler(executor, a2asrv.WithExtendedAgentCard(agentCard))

	// 1. JSON-RPC handler
	jsonrpcHandler := a2asrv.NewJSONRPCHandler(requestHandler)
	cardHandler := a2asrv.NewStaticAgentCardHandler(agentCard)

	mux := http.NewServeMux()
	mux.Handle(fmt.Sprintf("/jsonrpc%s", a2asrv.WellKnownAgentCardPath), cardHandler)
	mux.Handle("/jsonrpc", jsonrpcHandler)

	httpServer := &http.Server{
		Addr:              host + ":" + fmt.Sprintf("%d", *httpPort),
		Handler:           mux,
		ReadHeaderTimeout: 3 * time.Second,
	}

	// 2. gRPC handler
	grpcServer := grpc.NewServer()
	grpcHandler := a2agrpc.NewHandler(requestHandler)
	grpcHandler.RegisterWith(grpcServer)

	// 3. Run servers concurrently
	ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
	defer stop()

	g, ctx := errgroup.WithContext(ctx)

	g.Go(func() error {
		log.Printf("Starting HTTP server on %s:%d (JSON-RPC)", host, *httpPort)
		if err := httpServer.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			return fmt.Errorf("HTTP server failed: %w", err)
		}
		return nil
	})

	g.Go(func() error {
		lis, err := net.Listen("tcp", host+":"+fmt.Sprintf("%d", *grpcPort))
		if err != nil {
			return err
		}
		log.Printf("Starting gRPC server on %s:%d", host, *grpcPort)
		if err := grpcServer.Serve(lis); err != nil {
			return fmt.Errorf("gRPC server failed: %w", err)
		}
		return nil
	})

	g.Go(func() error {
		<-ctx.Done()
		log.Println("Shutting down servers...")

		shutdownCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()

		if err := httpServer.Shutdown(shutdownCtx); err != nil {
			log.Printf("HTTP server shutdown error: %v", err)
		}

		grpcServer.GracefulStop()
		log.Println("Servers closed.")
		return nil
	})

	return g.Wait()
}

File: itk/agents/python/v03/main.py
import asyncio  # noqa: I001
import base64
import logging
import signal
import uuid

from collections.abc import AsyncIterator, Callable
from typing import Any

import click
import grpc
import httpx
import uvicorn

from fastapi import FastAPI
from agents.python.v03.pyproto import instruction_pb2

from a2a.client import ClientConfig, ClientFactory
from a2a.grpc import a2a_pb2_grpc
from a2a.server.agent_execution import AgentExecutor, RequestContext
from a2a.server.apps import A2AFastAPIApplication, A2ARESTFastAPIApplication
from a2a.server.events import EventQueue
from a2a.server.request_handlers import DefaultRequestHandler, GrpcHandler
from a2a.server.tasks import InMemoryTaskStore, TaskUpdater
from a2a.types import (
    AgentCapabilities,
    AgentCard,
    AgentInterface,
    AgentSkill,
    FilePart,
    FileWithBytes,
    Message,
    Part,
    Role,
    TransportProtocol,
)
from a2a.utils import new_agent_text_message


logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def wrap_instruction_to_request(
    instruction: instruction_pb2.Instruction,
) -> Message:
    """Wraps an Instruction proto in an A2A Message.

    Args:
        instruction: The Instruction protobuf to wrap inside the message.

    Returns:
        Message: An A2A Message containing the serialized instruction as a byte file part.

    """
    inst_bytes = instruction.SerializeToString()
    b64_inst = base64.b64encode(inst_bytes).decode('utf-8')

    return Message(
        role=Role.user,
        message_id=str(uuid.uuid4()),
        parts=[
            Part(
                root=FilePart(
                    file=FileWithBytes(
                        bytes=b64_inst,
                        mime_type='application/x-protobuf',
                        name='instruction.bin',
                    )
                )
            )
        ],
        metadata={'a2a/protocol_version': '0.3'},
    )


async def get_client_with_transport(
    http_client: httpx.AsyncClient,
    url: str,
    transport: TransportProtocol | str,
    streaming: bool = False,
) -> Any:
    """Resolves the agent card and returns an A2AClient configured with the specified transport.

    Args:
        http_client: An asynchronous HTTPX client used for communication.
        url: The URL pointing to the agent's well-known card endpoint.
        transport: The requested transport protocol (e.g., 'jsonrpc', 'grpc', 'http_json').
        streaming: Whether to use streaming.

    Returns:
        Any: An initialized A2A client bound to the specified transport.

    Raises:
        ValueError: If the specified transport is not supported or recognized.

    """
    transport_map = {
        'jsonrpc': TransportProtocol.jsonrpc,
        'http_json': TransportProtocol.http_json,
        'grpc': TransportProtocol.grpc,
    }
    if not isinstance(transport, TransportProtocol):
        transport = transport_map.get(
            transport.lower() if isinstance(transport, str) else transport
        )

    if not transport:
        raise ValueError(f'Unsupported transport: {transport}')

    config = ClientConfig()
    config.httpx_client = http_client
    config.grpc_channel_factory = grpc.aio.insecure_channel
    config.supported_transports = [transport]
    config.use_client_preference = True
    config.streaming = streaming

    return await ClientFactory.connect(url, client_config=config)


async def handle_instruction(
    instruction: instruction_pb2.Instruction,
    call_agent_func: Callable[[instruction_pb2.CallAgent], AsyncIterator[str]],
) -> list[str]:
    """Processes a single Instruction proto recursively.

    Args:
        instruction: The starting Instruction protobuf.
        call_agent_func: An asynchronous callable that handles executing call_agent steps.

    Returns:
        list[str]: A list of string responses gathered from processing the instruction.

    Raises:
        ValueError: If the instruction type is neither call_agent, return_response, nor steps.

    """
    if instruction.HasField('call_agent'):
        return [p async for p in call_agent_func(instruction.call_agent)]
    if instruction.HasField('return_response'):
        return [instruction.return_response.response]
    if instruction.HasField('steps'):
        results = []
        for step in instruction.steps.instructions:
            results.extend(await handle_instruction(step, call_agent_func))
        return results
    raise ValueError('Unknown instruction type')


async def _call_agent_func(
    call_agent_proto: instruction_pb2.CallAgent,
) -> AsyncIterator[str]:
    logger.info(
        'Calling outbound agent: %s via %s',
        call_agent_proto.agent_card_uri,
        call_agent_proto.transport,
    )
    async with httpx.AsyncClient(timeout=30) as http_client:
        client = await get_client_with_transport(
            http_client,
            call_agent_proto.agent_card_uri,
            call_agent_proto.transport,
            streaming=call_agent_proto.streaming,
        )
        msg = wrap_instruction_to_request(call_agent_proto.instruction)
        async for event in client.send_message(msg):
            logger.info('Event received: %s: %s', type(event), event)

            message = None
            if hasattr(event, 'role') and hasattr(
                event, 'parts'
            ):  # Likely a Message
                message = event
            elif isinstance(event, tuple):
                for item in event:
                    if item is None:
                        continue
                    if hasattr(item, 'role') and hasattr(item, 'parts'):
                        message = item
                        break
                    status = getattr(item, 'status', None) or getattr(
                        getattr(item, 'status_update', None),
                        'status',
                        None,
                    )
                    if status and getattr(status, 'message', None):
                        message = status.message
                        break
                    if getattr(item, 'message', None) and hasattr(
                        item.message, 'parts'
                    ):
                        message = item.message
                        break

            if message:
                text_parts = []
                for p in message.parts:
                    p_root = getattr(p, 'root', p)
                    t = getattr(p_root, 'text', None)
                    if t:
                        text_parts.append(t)
                if text_parts:
                    yield '\n'.join(text_parts)


class V03AgentExecutor(AgentExecutor):
    """Simplified AgentExecutor for ITK v0.3 logic."""

    async def execute(
        self, context: RequestContext, event_queue: EventQueue
    ) -> None:
        """Implementation of the AgentExecutor interface.

        Args:
            context: The request context containing the incoming message.
            event_queue: The event queue to send responses to.

        """
        task_updater = TaskUpdater(
            event_queue,
            task_id=context.task_id or str(uuid.uuid4()),
            context_id=context.context_id or str(uuid.uuid4()),
        )
        await task_updater.submit()
        await task_updater.start_work()

        instruction = None
        # Extract proto from message parts
        for part in context.message.parts:
            part_root = part.root
            if isinstance(part_root, FilePart) and isinstance(
                part_root.file, FileWithBytes
            ):
                try:
                    raw_bytes = base64.b64decode(part_root.file.bytes)
                    instruction = instruction_pb2.Instruction()
                    instruction.ParseFromString(raw_bytes)
                    break
                except Exception:
                    logger.exception('Failed to parse Instruction proto')
                    continue

        if not instruction:
            logger.error('No valid Instruction found in message parts')
            error_msg = 'Error: No valid Instruction found in request.'
            await task_updater.failed(message=new_agent_text_message(error_msg))
            return

        try:
            result = await handle_instruction(instruction, _call_agent_func)
            response_msg = new_agent_text_message('\n'.join(result))
            await task_updater.complete(message=response_msg)
        except Exception:
            logger.exception('Instruction execution failed')
            error_msg = 'Execution Error: instruction handling failed'
            await task_updater.failed(message=new_agent_text_message(error_msg))

    async def cancel(
        self, context: RequestContext, event_queue: EventQueue
    ) -> None:
        """Implementation of cancel request method required by AgentExecutor interface.

        Args:
            context: The request context containing the incoming message.
            event_queue: The event queue to send responses to.

        """


async def create_grpc_server(
    agent_card: AgentCard,
    request_handler: DefaultRequestHandler,
    host: str,
    port: int,
) -> grpc.aio.Server:
    """Creates and configures the gRPC server.

    Args:
        agent_card: The AgentCard specifying the server's identity and capabilities.
        request_handler: The request handler used for routing and handling instructions.
        host: The host address to bind the server to (e.g., '127.0.0.1').
        port: The port to bind the gRPC server.

    Returns:
        grpc.aio.Server: An initialized gRPC Server object.

    """
    server = grpc.aio.server()
    a2a_pb2_grpc.add_A2AServiceServicer_to_server(
        GrpcHandler(agent_card, request_handler), server
    )
    server.add_insecure_port(f'{host}:{port}')
    return server


def create_http_server(
    agent_card: AgentCard,
    request_handler: DefaultRequestHandler,
    host: str,
    port: int,
) -> uvicorn.Server:
    """Creates and configures the HTTP server for JSON-RPC and REST via FastAPI.

    Args:
        agent_card: The AgentCard specifying the server's identity and capabilities.
        request_handler: The request handler used for routing and handling instructions.
        host: The host address to bind the server to (e.g., '127.0.0.1').
        port: The port to bind the HTTP server.

    Returns:
        uvicorn.Server: An initialized Uvicorn Server instance holding the FastAPI app.

    """
    app = FastAPI(title='ITK v03 Agent Server (Consolidated)')

    app.mount(
        '/jsonrpc', A2AFastAPIApplication(agent_card, request_handler).build()
    )
    app.mount(
        '/rest', A2ARESTFastAPIApplication(agent_card, request_handler).build()
    )
    return uvicorn.Server(
        uvicorn.Config(app, host=host, port=port, log_config=None)
    )


async def _run_agent(http_port: int, grpc_port: int) -> None:
    host = '127.0.0.1'

    skill = AgentSkill(
        id='itk_v03_proto_skill',
        name='ITK v03 Proto Skill',
        description='Handles raw byte Instruction protos in v03 subproject.',
        tags=['proto', 'v03', 'itk'],
        examples=['Roll a dice', 'Call another agent'],
    )

    agent_card = AgentCard(
        name='ITK v03 Agent',
        description='Multi-transport agent supporting raw Instruction protos (Consolidated).',
        url=f'http://{host}:{http_port}/jsonrpc/',
        version='0.3.0',
        protocol_version='0.3.0',
        default_input_modes=['text'],
        default_output_modes=['text'],
        capabilities=AgentCapabilities(streaming=True),
        skills=[skill],
        preferred_transport=TransportProtocol.jsonrpc,
        additional_interfaces=[
            AgentInterface(
                url=f'http://{host}:{http_port}/rest/',
                transport=TransportProtocol.http_json,
            ),
            AgentInterface(
                url=f'{host}:{grpc_port}',
                transport=TransportProtocol.grpc,
            ),
        ],
    )

    request_handler = DefaultRequestHandler(
        agent_executor=V03AgentExecutor(), task_store=InMemoryTaskStore()
    )

    http_server = create_http_server(
        agent_card, request_handler, host, http_port
    )
    grpc_server = await create_grpc_server(
        agent_card, request_handler, host, grpc_port
    )

    # Signal handling
    loop = asyncio.get_running_loop()

    async def shutdown() -> None:
        logger.info('Shutting down...')
        http_server.should_exit = True
        await grpc_server.stop(5)

    for sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(sig, lambda: asyncio.create_task(shutdown()))

    await grpc_server.start()
    await http_server.serve()


@click.command()
@click.option('--httpPort', 'http_port', default=10101)
@click.option('--grpcPort', 'grpc_port', default=11001)
def main(http_port: int, grpc_port: int) -> None:
    """Command line entry point for starting the ITK v03 merged Agent.

    Args:
        http_port: The HTTP port to listen on for REST/JSON-RPC calls.
        grpc_port: The gRPC port to listen on.

    """
    asyncio.run(_run_agent(http_port, grpc_port))


if __name__ == '__main__':
    main()
